### PR TITLE
Implement API and CRD for ztunnel component

### PR DESCRIPTION
**Pull request by yxun**
- [X] Enhancement / New Feature
- [ ] Bug Fix
- [ ] Refactor
- [ ] Optimization
- [ ] Test
- [ ] Documentation Update

#### What this PR does / why we need it:
Hello, this is an initial draft of installing Istio Ambient ztunnel component using Sail Operator.
This PR includes a Sail Enhancement Proposal (SEP) (SEP3-ambient-mode.md) and inital CRD and API implementations.

We want to have an early review of the SEP3 proposal and we are working on new controller implementations next.

Related Issue/PR #
https://issues.redhat.com/browse/OSSM-8381

### CODE REVIEW

File: enhancements/SEP3-ambient-mode.md [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+
+## Goals
+
+- Provide a way for installing the ambient mode component profile and charts in Sail Operator. This includes the ztunnel component chart and binary, Istio CNI changes and related meshConfig in pilot.
+- Provide a way for managing the ambient mode components life cycle. The implementation should enable in-place upgrades.
+
+## Non-goals
+
+- Uninstallation of only the ambient mode components. Currently, to cleanup the environment, need to compeletely uninstall Istio from a cluster.
+- Istio Sidecar mode migration to ambient mode.
```
Comment (sridhargaddam) [inline enhancements/SEP3-ambient-mode.md:21]: What about the following sentence?

Migration of Istiod control plane and dataplane components between sidecar mode and Ambient dataplane mode (and vice versa)

**Review by sridhargaddam — COMMENTED** [review]
Thanks @yxun for the PR.

File: enhancements/SEP3-ambient-mode.md [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+
+In the `spec.values` field, users can specify helm chart values for the ztunnel component. These APIs are mapped from upstream Istio `manifests/charts/ztunnel/values.yaml` file. Users can also specify global values in the `spec.values.global` field. These APIs are the same as the IstioCNI global values APIs.
+
+#### IstioZtunnel Status
+
+The `status.state` field gives a quick hint as to whether a ztunnel DaemonSet has been reconcile ("Healthy") or if there are any problems with it.
+
+### Architecture
+
+To provide a developer preview of Istio Ambient mode using Sail Operator, we will need to design and implement new Sail Operator APIs, CRD and controller(s) for installing Istio Ambient mode profile and helm charts. This involves the following component helm charts and values:
```
Comment (sridhargaddam) [inline enhancements/SEP3-ambient-mode.md:67]: IMO `developer preview` is more of a product support stance. Here we can say...

```suggestion
To provide support for Istio Ambient mode using Sail Operator, we will need to design and implement new Sail Operator APIs, CRD and controller(s) for installing Ambient profile. This involves the following changes:
```

File: enhancements/SEP3-ambient-mode.md [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+* Set Istio `meshConfig.defaultConfig.proxyMetadata.ISTIO_META_ENABLE_HBONE: "true"`
+
+Therefore, we'll start from introducing a new Sail Operator CRD called `IstioZtunnel` and mapping ztunnel helm chart values to the CRD APIs.
+
+### Performance Impact
+
+### Backward Compatibility
+
+### Kubernetes vs OpenShift vs Other Distributions
+
```
[the remark this answers is not in the document]
Comment (sridhargaddam) [reply enhancements/SEP3-ambient-mode.md:81]: Just curious: ztunnel is only applicable in Ambient mode, so do we really require users to specify `profile` (as platform details are now separated)?

[the remark this answers is not in the document]
Comment (luksa) [reply enhancements/SEP3-ambient-mode.md:81]: You're right. The `profile` field in `ZTunnel` should default to `ambient`. Users don't need to specify it, but they can, if they want to use another profile (not applicable yet, but eventually we'll probably support user-provided profiles just like upstream).

**Comment by yxun** [comment]
Currently, ZTunnelConfig and ZTunnelGlobalConfig structs are defined in the types_extra.go.  
I am working on an upstream Istio change about those two. After that, I will remove those two in the Sail Operator API types_extra.go and then our transformer tool can pull those from upstream to types.gen.go.

File: api/v1alpha1/ztunnel_types.go [hunk 4]
```diff
@@ -0,0 +1,198 @@
...
+	// Defines the version of Istio to install.
+	// Must be one of: v1.24.0 or latest.
+	// +operator-sdk:csv:customresourcedefinitions:type=spec,order=1,displayName="Istio Version",xDescriptors={"urn:alm:descriptor:com.tectonic.ui:fieldGroup:General", "urn:alm:descriptor:com.tectonic.ui:select:v1.24.0", "urn:alm:descriptor:com.tectonic.ui:select:latest"}
+	// +kubebuilder:validation:Enum=v1.24.0;latest
+	// +kubebuilder:default=v1.24.0
+	Version string `json:"version"`
+
+	// +sail:profile
+	// The built-in installation configuration profile to use.
+	// The 'default' profile is always applied. On OpenShift, the 'openshift' profile is also applied on top of 'default'.
```
Comment (luksa) [inline api/v1alpha1/ztunnel_types.go:40]: This should say that the `ambient` profile is the default. Also, the `openshift` profile is no longer used.

File: api/v1alpha1/ztunnel_types.go [hunk 5]
```diff
@@ -0,0 +1,198 @@
...
+	// Must be one of: ambient, default, demo, empty, external, openshift-ambient, openshift, preview, remote, stable.
+	// +++PROFILES-DROPDOWN-HIDDEN-UNTIL-WE-FULLY-IMPLEMENT-THEM+++operator-sdk:csv:customresourcedefinitions:type=spec,displayName="Profile",xDescriptors={"urn:alm:descriptor:com.tectonic.ui:fieldGroup:General", "urn:alm:descriptor:com.tectonic.ui:select:ambient", "urn:alm:descriptor:com.tectonic.ui:select:default", "urn:alm:descriptor:com.tectonic.ui:select:demo", "urn:alm:descriptor:com.tectonic.ui:select:empty", "urn:alm:descriptor:com.tectonic.ui:select:external", "urn:alm:descriptor:com.tectonic.ui:select:minimal", "urn:alm:descriptor:com.tectonic.ui:select:preview", "urn:alm:descriptor:com.tectonic.ui:select:remote"}
+	// +operator-sdk:csv:customresourcedefinitions:type=spec,xDescriptors={"urn:alm:descriptor:com.tectonic.ui:hidden"}
+	// +kubebuilder:validation:Enum=ambient;default;demo;empty;external;openshift-ambient;openshift;preview;remote;stable
+	Profile string `json:"profile,omitempty"`
```
Comment (luksa) [inline api/v1alpha1/ztunnel_types.go:45]:
```suggestion
	// +kubebuilder:default=ambient
	Profile string `json:"profile,omitempty"`
```

File: api/v1alpha1/ztunnel_types.go [hunk 6]
```diff
@@ -0,0 +1,198 @@
...
+// +kubebuilder:printcolumn:name="Version",type="string",JSONPath=".spec.version",description="The version of the Istio ztunnel installation."
+// +kubebuilder:printcolumn:name="Age",type="date",JSONPath=".metadata.creationTimestamp",description="The age of the object"
+// +kubebuilder:validation:XValidation:rule="self.metadata.name == 'default'",message="metadata.name must be 'default'"
+
+// ZTunnel represents a deployment of the Istio ztunnel component.
+type ZTunnel struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	// +kubebuilder:default={version: "v1.24.0", namespace: "kube-system"}
```
Comment (luksa) [inline api/v1alpha1/ztunnel_types.go:181]:
```suggestion
	// +kubebuilder:default={version: "v1.24.0", namespace: "kube-system", profile: "ambient"}
```

File: api/v1alpha1/ztunnel_types.go [hunk 7]
```diff
@@ -0,0 +1,198 @@
...
+	// +++PROFILES-DROPDOWN-HIDDEN-UNTIL-WE-FULLY-IMPLEMENT-THEM+++operator-sdk:csv:customresourcedefinitions:type=spec,displayName="Profile",xDescriptors={"urn:alm:descriptor:com.tectonic.ui:fieldGroup:General", "urn:alm:descriptor:com.tectonic.ui:select:ambient", "urn:alm:descriptor:com.tectonic.ui:select:default", "urn:alm:descriptor:com.tectonic.ui:select:demo", "urn:alm:descriptor:com.tectonic.ui:select:empty", "urn:alm:descriptor:com.tectonic.ui:select:external", "urn:alm:descriptor:com.tectonic.ui:select:minimal", "urn:alm:descriptor:com.tectonic.ui:select:preview", "urn:alm:descriptor:com.tectonic.ui:select:remote"}
+	// +operator-sdk:csv:customresourcedefinitions:type=spec,xDescriptors={"urn:alm:descriptor:com.tectonic.ui:hidden"}
+	// +kubebuilder:validation:Enum=ambient;default;demo;empty;external;openshift-ambient;openshift;preview;remote;stable
+	Profile string `json:"profile,omitempty"`
+
+	// Namespace to which the Istio ztunnel component should be installed.
+	// +operator-sdk:csv:customresourcedefinitions:type=spec,xDescriptors={"urn:alm:descriptor:io.kubernetes:Namespace"}
+	// +kubebuilder:default=kube-system
```
Comment (luksa) [inline api/v1alpha1/ztunnel_types.go:49]: Is this correct? Do we want to deploy ztunnel to `kube-system`? For `IstioCNI` we use `istio-cni` instead of `kube-system`.

**Review by luksa — COMMENTED** [review]
A few comments. Also, please add `ZTunnel` to the `PROJECT` file. Without this, some gen scripts don't work properly.

File: api/v1alpha1/ztunnel_types.go (hunk 7 above)
Comment (sridhargaddam) [reply api/v1alpha1/ztunnel_types.go:49]: If we change the namespace to `istio-cni`, we will also have to modify the trustedZtunnelNamespace which is configured to `kube-system`.

Comment (luksa) [reply api/v1alpha1/ztunnel_types.go:49]: I don't think it's the best idea to use `kube-system` on OpenShift, as all the other OpenShift components intentionally typically use other namespaces, never `kube-system`.

Comment (sridhargaddam) [reply api/v1alpha1/ztunnel_types.go:49]: Agree with Marko. In that case, we should probably update the upstream charts/docs as well to use `istio-cni` namespace instead of `kube-system`. 

@yxun when you tried installing ztunnel in `istio-cni`, did you also configure `trustedZtunnelNamespace` to `istio-cni`? If so, did it work or did you notice any additional errors?

Comment (sridhargaddam) [reply api/v1alpha1/ztunnel_types.go:49]: > I don't think it's the best idea to use `kube-system` on OpenShift, as all the other OpenShift components intentionally use other namespaces, never `kube-system`.

@luksa, I see that when using istioctl with `openshift` profile (i.e., non ambient use-case) the `istio-cni` pods are installed in the `kube-system` namespace itself and not in `istio-cni` namespace.

Comment (yxun) [reply api/v1alpha1/ztunnel_types.go]: I updated this default namespace value in the `ztunnel_types.go` file.
The Helm chart profile values need to be updated from upstream if we want all openshift related profiles use `trustedZtunnelNamespace: istio-cni`

Comment (sridhargaddam) [reply api/v1alpha1/ztunnel_types.go]: As discussed during the call yesterday, we will be installing the ztunnel CR in the `ztunnel` namespace and not in `istio-cni`. I will take care of trustedZtunnelNamespace validation/modification as part of my PR.

File: enhancements/SEP3-ambient-mode.md
Comment (luksa) [inline enhancements/SEP3-ambient-mode.md:1]: I think we talked about moving this to a separate PR last week.

File: enhancements/SEP3-ambient-mode.md [hunk 8]
```diff
@@ -0,0 +1,93 @@
...
+  namespace: kube-system
+spec:
+    name: default
+    version: v1.24.0    # Defines the version of Istio
+    profile: ambient    # The built-in installation configuration profile to use.
+    values:
+      ztunnel:
+        hub: gcr.io/istio-testing
+        tag: latest
+        image: ztunnel
```
Comment (luksa) [inline enhancements/SEP3-ambient-mode.md:50]: Is this required?

File: enhancements/SEP3-ambient-mode.md [hunk 9]
```diff
@@ -0,0 +1,93 @@
...
+### Performance Impact
+
+### Backward Compatibility
+
+### Kubernetes vs OpenShift vs Other Distributions
+
+## Alternatives Considered
+
+### Reuse `IstioCNI`'s type field for specifying a ztunnel component
+We could expend the `IstioCNI` resource APIs and `IstioCNISpec` values field for a ztunnel component installation and life-cycle management. It would have the benefit that combines IstioCNI and ztunnel components together because both of them are deployed as a Kubernetes DaemonSet on each node. The disadvantage though is that it would be confusing for users to manage either IstioCNI or ztunnel life-cycle.
```
Comment (luksa) [inline enhancements/SEP3-ambient-mode.md:85]: Didn't we also consider a several other options? At least the option of having ZTunnel deployed via the Istio/IstioRevision resource. It would be great to explain here why we decided against these alternative options (this is the entire point of this section).

Comment (yxun) [reply enhancements/SEP3-ambient-mode.md:85]: I see. I will update that alternative options section in SEP3 and then we can discuss this proposal.

File: enhancements/SEP3-ambient-mode.md [hunk 10]
```diff
@@ -0,0 +1,93 @@
...
+spec:
+    name: default
+    version: v1.24.0    # Defines the version of Istio
+    profile: ambient    # The built-in installation configuration profile to use.
+    values:
+      ztunnel:
+        hub: gcr.io/istio-testing
+        tag: latest
+        image: ztunnel
+        istioNamespace: istio-system
```
[the remark this answers is not in the document]
Comment (sridhargaddam) [reply enhancements/SEP3-ambient-mode.md:51]: In an Ambient deployment, the ztunnel ds pod requests the Istiod control plane to issue certificates for itself and for the workloads (on the same node) that it manages. So, it needs the information of CA_ADDRESS and XDS_ADDRESS. The lifecycle of ztunnel is different from the Istiod, but it needs details of the istiod control plane. 

```
xdsAddress: By default, it is istiod.istio-system.svc:15012 if revision is not set, or istiod-<revision>.<istioNamespace>.svc:15012
```
For more details, see : https://github.com/istio/istio/blob/master/manifests/charts/ztunnel/values.yaml#L79-L89
